articles/nginx-reverse-proxy/index.md

```diff
@@ -0,0 +1,161 @@
+Another example could be a particular route like domain/client and domain/server. The only condition for the distinguishing element is to follow a Valid URL Regular Expression. Learn about Regex from [here](https://regexr.com/).
+
+
+## Step 2: Add DNS records
+This is the part where one would add the DNS records in their DNS management dashboard. If you are running Nginx locally, you can skip this step.
+
+The general DNS Configurations would be something like
+- Server app mapped to the server.domain
+- Client app mapped to the client.domain
+
+My Localhost Config, in this case, would be
+- Server mapped to server.localhost
+- Client mapped to client.localhost
+
+There are two standard protocols HTTP and HTTPS. The default port for HTTP is 80 and HTTPS is 443. This is the reason we must not run our applications on these ports because our NGINX server would be running on these two ports. All the requests the client makes would either be redirected to Port 80 or 443 from where it would be redirected internally to the corresponding application.
+
+## Step 4 - Configure NGINX at 80 for HTTP and 443 for HTTPS
+Now that we have our apps running and our DNS records ready. We can start configuring our NGINX Reverse Proxy to make it all work.
+Navigate into the Nginx folder and locate  ```nginx.conf``` .
+
+Add these configurations inside the HTTP block.
+
+### Step 4.1 - HTTP
+
+```
+server {       
+     listen       80;
+     server_name  server.domain;
+     location / {
+          proxy_pass "http://localhost:3000" ;
+     }
+}
+server {       
+     listen       80;
+     server_name  client.domain;
+
+     location / {
+          proxy_pass "http://localhost:3001" ;
+     }
+}
+```
+
+### Step 4.2 - HTTPS
+
+```
+server{
+    listen 443 ssl http2;
+    listen [::]:443 ssl http2;
+
+    server_name server.domain;
+
+    location / {    
+        proxy_pass "http://localhost:3000/";
+    }
+
+    ssl_certificate <location of SSL certificate>
+    ssl_certificate_key <location of SSL certificate Key>
+}
+server{
+    listen 443 ssl http2;
+    listen [::]:443 ssl http2;
+
+    server_name client.domain;
+
+    location / {    
+        proxy_pass "http://localhost:3001/";
+    }
+
+    ssl_certificate <location of SSL certificate>
+    ssl_certificate_key <location of SSL certificate Key>
+}
+```
+
+**NOTE** 
+These are the basic minimum configurations required to successfully configure NGINX for reverse proxying. Feel free to explore other config parameters as well. 
+
+Change the domain name to your domain.
+For a SSL Certificate and Key, you can obtain them from your SSL provider. If you don't have one, Use this free service [LetsEncrypt](https://letsencrypt.org/). Follow their documentation to get Free SSL instantly!
```

Comment:
Tried to run certbot for Let's Encrypt on Debian and received `Error creating new order :: Cannot issue for "client.domain": Domain name does not end with a valid public suffix (TLD)` error. Should this be changed to client.localhost in a localhost environment?

Reply:
A valid TLD(Top level domain) would be something like ```section.io``` . So when you run certbot you need to give a valid domain. This was the cause of the error. And yes you need to change it to ```client.localhost``` if you are trying to get SSL for your localhost environment.